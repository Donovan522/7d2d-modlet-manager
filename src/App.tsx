import Box from "@material-ui/core/Box";
import Button from "@material-ui/core/Button";
import CircularProgress from "@material-ui/core/CircularProgress";
import Collapse from "@material-ui/core/Collapse";
import Container from "@material-ui/core/Container";
import CssBaseline from "@material-ui/core/CssBaseline";
import Fab from "@material-ui/core/Fab";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import List from "@material-ui/core/List";
import { makeStyles, ThemeProvider } from "@material-ui/core/styles";
import Switch from "@material-ui/core/Switch";
import LaunchIcon from "@material-ui/icons/Launch";
import { execFile } from "child_process";
import { remote } from "electron";
import FolderPicker from "src/components/FolderPicker";
import Modlets from "src/components/Modlets";
import { fileExists, GameXML, getModlets } from "src/helpers";
import theme from "src/helpers/theme";
import menuTemplate from "./menu";
import path from "path";
import React, { useCallback, useEffect, useReducer, useState } from "react";
import { hot } from "react-hot-loader";

const useStyles = makeStyles(theme => ({
  mainContainer: {
    display: "flex",
    flexDirection: "column",
    height: "100vh"
  },
  bodyContainer: {
    marginTop: 10,
    overflowY: "auto",
    height: "100%"
  },
  controlsContainer: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    alignContent: "center"
  },
  modeControl: {
    padding: theme.spacing(1)
  },
  noGameFolder: {
    display: "block",
    margin: "auto",
    marginTop: 200
  },
  launchButton: {
    margin: theme.spacing(1)
  },
  launchIcon: {
    marginRight: theme.spacing(1),
    color: theme.palette.secondary.main
  }
}));

interface AppProps {
  store: any;
}

function App(props: AppProps): React.ReactElement {
  const classes = useStyles();
  const gameExecutable = "7DaysToDie.exe";
  // const gameExecutableEAC = "7DaysToDie_EAC.exe";
  const [loading, setLoading] = useState(false);

  const initialState: IState = {
    advancedMode: false,
    config: {
      gameFolder: "",
      modletFolder: "",
      mode: 0
    },
    gameXML: null,
    modlets: []
  };

  const initState = (): IState => ({
    advancedMode: !!parseInt(props.store.get("mode")),
    config: props.store.store,
    gameXML: props.store.get("gameFolder") ? new GameXML(props.store.get("gameFolder")) : null,
    modlets: []
  });

  const sortModlets = (a: IModletState, b: IModletState) => (a.modlet.get("name") > b.modlet.get("name") ? 1 : -1);

  const stateReducer = (state: IState, action: { type: string; payload?: any }): IState => {
    switch (action.type) {
      case "setAdvancedMode": {
        props.store.set("mode", action.payload ? 1 : 0);

        return {
          ...state,
          advancedMode: action.payload
        };
      }

      case "setGameFolder": {
        props.store.set("gameFolder", action.payload);

        return {
          ...state,
          config: { ...props.store.store, gameFolder: action.payload },
          gameXML: new GameXML(action.payload)
        };
      }

      case "setModletFolder": {
        props.store.set("modletFolder", action.payload);

        return {
          ...state,
          config: { ...props.store.store, modletFolder: action.payload },
          modlets: getModlets(action.payload)
        };
      }

      case "setModlets": {
        return {
          ...state,
          modlets: action.payload.sort(sortModlets)
        };
      }

      case "clearModlets": {
        return {
          ...state,
          modlets: []
        };
      }

      case "syncModlets": {
        const modletState = state.modlets.filter((obj: IModletState) => obj.modlet === action.payload.modlet)[0];

        if (!modletState) throw new Error("syncModlets dispatch received invalid modlet");

        const modletStates = state.modlets.filter((obj: IModletState) => obj.modlet !== action.payload.modlet);
        const newModletState = { ...modletState, ...action.payload };

        return {
          ...state,
          modlets: [...modletStates, newModletState].sort(sortModlets)
        };
      }

      case "setGameXML": {
        return {
          ...state,
          gameXML: new GameXML(state.config.gameFolder)
        };
      }

      default:
        return state;
    }
  };

  const [state, stateDispatch] = useReducer(stateReducer, initialState, initState);

  const errorDialog = (title: string, err: Error) => {
    remote.dialog.showMessageBox({
      type: "error",
      title: title,
      message: err.message
    });
  };

  const launchGame = () => {
    const game = path.normalize(path.join(state.config.gameFolder, gameExecutable));

    if (fileExists(game))
      execFile(game, err => {
        if (err) {
          errorDialog("Unable to start 7 Days to Die", err);
          return;
        }
      });
  };

  const refreshModlets = useCallback(
    (modletsPath?: string) => {
      let newModletList = getModlets(
        modletsPath || state.advancedMode ? state.config.modletFolder : path.join(state.config.gameFolder, "Mods")
      );
      if (newModletList.length) stateDispatch({ type: "setModlets", payload: newModletList });
    },
    [state.advancedMode, state.config.gameFolder, state.config.modletFolder]
  );

  const getFolder = async (title: string) => {
    setLoading(true);

    const dialog = await remote.dialog.showOpenDialog({
      title: title,
      properties: ["openDirectory"]
    });

    setLoading(false);

    if (!dialog.canceled) return dialog.filePaths[0];

    return null;
  };

  const getGameFolder = useCallback(() => {
    getFolder('Please select the "7 Days to Die" game folder').then(newFolder => {
      if (newFolder) {
        if (fileExists(path.join(newFolder, gameExecutable))) {
          stateDispatch({ type: "setGameFolder", payload: newFolder });
          if (!state.config.modletFolder) {
            const newModletFolder = path.posix.join(newFolder, "Mods");
            stateDispatch({ type: "setModletFolder", payload: newModletFolder });
          } else {
            refreshModlets();
          }
        } else {
          errorDialog("Invalid Game Folder", new Error("The chosen folder is not a valid 7 Days to Die game folder"));
        }
      }
    });
  }, [refreshModlets, state.config.modletFolder]);

  const getModletFolder = () => {
    getFolder('Please Select a valid "7 Days to Die" Modlet Folder').then(newFolder => {
      if (newFolder) {
        stateDispatch({ type: "setModletFolder", payload: newFolder });
      }
    });
  };

  const toggleAdvancedMode = () => {
    stateDispatch({ type: "clearModlets" });
    stateDispatch({ type: "setAdvancedMode", payload: !state.advancedMode });
  };

  useEffect(() => {
    if (!state.config.gameFolder) getGameFolder();
  }, [state.config.gameFolder, getGameFolder]);

  useEffect(() => {
    if (state.config.mode === undefined) stateDispatch({ type: "setAdvancedMode", payload: false });
  }, [state.config.mode]);

  useEffect(() => {
    if (!loading && !state.config.modletFolder && state.config.gameFolder) {
      const newModletFolder = path.join(state.config.gameFolder, "Mods");
      stateDispatch({ type: "setModletFolder", payload: newModletFolder });
    }
  }, [loading, state.config.gameFolder, state.config.modletFolder]);

  useEffect(() => {
    if (state.config.modletFolder && state.config.gameFolder) refreshModlets();
  }, [refreshModlets, state.config.modletFolder, state.config.gameFolder]);

  if (!loading && (state === undefined || state.config.gameFolder === undefined)) {
    return (
      <Button variant="contained" color="secondary" onClick={getGameFolder} className={classes.noGameFolder}>
        Please Select the 7 days to die game folder to continue...
      </Button>
    );
  }

  if (loading && (!state.config.gameFolder || !state.config.modletFolder))
    return <CircularProgress className={classes.noGameFolder} />;

  const commands = {
    chooseGameFolder: getGameFolder,
    chooseModletFolder: getModletFolder,
    toggleMode: toggleAdvancedMode,
    refreshModlets: refreshModlets
  };

  // @ts-ignore
  remote.Menu.setApplicationMenu(remote.Menu.buildFromTemplate(menuTemplate(commands)));

  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <Box className={classes.mainContainer}>
        <Container maxWidth="xl" className={classes.bodyContainer}>
          <Box className={classes.controlsContainer}>
            <FormControlLabel
              className={classes.modeControl}
              control={<Switch size="small" checked={state.advancedMode} onChange={toggleAdvancedMode} />}
              label={state.advancedMode ? "Advanced Mode" : "Basic Mode"}
            />
            <Fab
              variant="extended"
              size="medium"
              color="primary"
              aria-label="Launch 7 Days to Die"
              className={classes.launchButton}
              onClick={launchGame}
            >
              <LaunchIcon className={classes.launchIcon} />
              Play Game
            </Fab>
          </Box>
          <List dense={true}>
            <FolderPicker
              advancedMode={state.advancedMode}
              folder={state.config.gameFolder}
              handleClick={getGameFolder}
              label="Game Folder"
              tooltip="Click to select Game Folder"
            />
            <Collapse in={state.advancedMode}>
              <FolderPicker
                advancedMode={state.advancedMode}
                folder={state.config.modletFolder}
                handleClick={getModletFolder}
                label="Modlet Folder"
                tooltip="Click to select Modlet folder"
              />
            </Collapse>
          </List>
          <Modlets state={state} stateDispatch={stateDispatch} />
        </Container>
      </Box>
    </ThemeProvider>
  );
}

export default hot(module)(App);
