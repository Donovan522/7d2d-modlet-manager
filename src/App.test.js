import { shallow } from "enzyme";
import App from "App";
import React from "react";
import mock_store from "test_helpers/mock_store";

jest.mock("electron", () => ({
  remote: {
    Menu: {
      buildFromTemplate: jest.fn(),
      setApplicationMenu: jest.fn()
    }
  }
}));

jest.mock("electron-util", () => ({
  aboutMenuItem: jest.fn()
}));

it("renders without crashing", () => {
  shallow(<App store={mock_store} />);
});
