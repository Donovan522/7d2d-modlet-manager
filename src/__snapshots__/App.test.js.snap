// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-10 makeStyles-mainContainer-1"
  >
    <div
      class="MuiContainer-root makeStyles-bodyContainer-2 MuiContainer-maxWidthXl"
    >
      <div
        class="MuiBox-root MuiBox-root-19 makeStyles-controlsContainer-3"
      >
        <label
          class="MuiFormControlLabel-root makeStyles-modeControl-4"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeSmall"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-38 MuiSwitch-switchBase MuiSwitch-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-41 MuiSwitch-input"
                  type="checkbox"
                  value=""
                />
                <span
                  class="MuiSwitch-thumb"
                />
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Basic Mode
          </span>
        </label>
        <button
          aria-label="Launch 7 Days to Die"
          class="MuiButtonBase-root MuiFab-root makeStyles-launchButton-7 MuiFab-extended MuiFab-sizeMedium MuiFab-primary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiFab-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-launchIcon-8"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
              />
            </svg>
            Play Game
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <ul
        class="MuiList-root MuiList-dense MuiList-padding"
      >
        <li
          class="MuiListItem-root makeStyles-listItem-107 MuiListItem-dense MuiListItem-gutters"
          color="secondary"
          title="Click to select Game Folder"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
            >
              Game Folder
            </span>
            <p
              class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
            >
              /foo/bar
            </p>
          </div>
        </li>
        <div
          class="MuiCollapse-container MuiCollapse-hidden"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner"
            >
              <li
                class="MuiListItem-root makeStyles-listItem-107 MuiListItem-dense MuiListItem-gutters"
                color="secondary"
                title="Click to select Modlet folder"
              >
                <div
                  class="MuiListItemIcon-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
                  >
                    Modlet Folder
                  </span>
                  <p
                    class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
                  >
                    /foo/bar/Mods
                  </p>
                </div>
              </li>
            </div>
          </div>
        </div>
      </ul>
      <div
        class="MuiGrid-root makeStyles-root-149 MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root makeStyles-cardEmpty-145 MuiGrid-item"
        >
          <h6
            class="MuiTypography-root makeStyles-noModletsHeader-146 MuiTypography-h6"
          >
            No Modlets Installed
          </h6>
          <hr
            class="MuiDivider-root MuiDivider-middle"
          />
          <p
            class="MuiTypography-root makeStyles-noModletsBody-147 MuiTypography-body1"
          >
            To install mods, place them in \\foo\\bar\\Mods
            <span />
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
